### PR TITLE
KokkosKernels: match #2353

### PR DESCRIPTION
**Pull request by cwpearson**
@trilinos/kokkos-kernels

Matching patch for https://github.com/kokkos/kokkos-kernels/pull/2353/

A performance regression in `KokkosSparse::sort_crs_matrix` was introduced in https://github.com/trilinos/Trilinos/pull/13401, where `sort_crs_matrix` was changed to use `Kokkos::Experimental::sort_by_key` in certain cases. 
One of the MueLu setup performance tests catches one of those cases:

```bash
muelu/test/scaling/MueLu_Driver.exe" "--stacked-timer" "--nx=120" "--ny=120" "--nz=120" "--matrixType=Laplace3D"
```

On MI300 this manifested as a ~60x slowdown in Tpetra's matrix transpose because `sort_by_key` is only fast on AMD when Kokkos can use the rocThrust TPL, which Trilinos does not support. This PR also preemptively attempts to avoid falling in to the same trap on SYCL. For CUDA, Thrust TPL is always available, so this is not a problem.

| Config | Local Transpose Time (s) |
|-|-|
| Before 13401 | 0.056 |
| After 13401 | 0.34 |
| This PR | 0.056 |

### CODE REVIEW

**Review by ndellingwood — APPROVED** [review]
Changes in https://github.com/kokkos/kokkos-kernels/pull/2353 were approved and AT2 testing passed
Please hold off on merge of this PR until the corresponding kokkos-kernels PR merges to develop

**Comment by ndellingwood** [comment]
@cwpearson in case it is helpful as a follow up, I had started a PR https://github.com/trilinos/Trilinos/pull/12826 to add rocthrust support in Trilinos, if useful that can be revisited

**Comment by cwpearson** [comment]
Let's revive that effort!

**Comment by masterleinad** [comment]
> `sort_by_key` is only fast on AMD when Kokkos can use the rocThrust TPL, which Trilinos does not support.

I would expect that you can enable `rocThrust` in `Trilinos` via `Kokkos` after https://github.com/kokkos/kokkos/pull/6164. I'd be curious to hear about problems with that.

**Comment by cwpearson** [comment]
Ah okay, we're just waiting for that change to make its way to `develop` (I guess one could do an external kokkos build today)

**Comment by cwpearson** [comment]
I think this PR will coexist peacefully with that change.

**Comment by masterleinad** [comment]
> I think this PR will coexist peacefully with that change - you can see the linked Kokkos Kernels issue, but even enabling rocThrust is not necessarily a win (though it is an improvement over sorting on the host 😄 )

Yes, I just wanted to clarify.

**Review by lucbv — APPROVED** [review]
Lgtm

**Pull request merged**
